includes/Core/Util/Debug_Data.php

```diff
@@ -0,0 +1,272 @@
+<?php
+/**
+ * Class Google\Site_Kit\Core\Util\Debug_Data
+ *
+ * @package   Google\Site_Kit\Core\Util
+ * @copyright 2020 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\Util;
+
+use Google\Site_Kit\Context;
+use Google\Site_Kit\Core\Authentication\Authentication;
+use Google\Site_Kit\Core\Authentication\Clients\OAuth_Client;
+use Google\Site_Kit\Core\Authentication\Credentials;
+use Google\Site_Kit\Core\Modules\Module;
+use Google\Site_Kit\Core\Modules\Module_With_Debug_Fields;
+use Google\Site_Kit\Core\Modules\Modules;
+use Google\Site_Kit\Core\Storage\Encrypted_Options;
+use Google\Site_Kit\Core\Storage\Options;
+use Google\Site_Kit\Core\Storage\User_Options;
+
+/**
+ * Class for integrating debug information with Site Health.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+class Debug_Data {
+	/**
+	 * Context instance.
+	 *
+	 * @since n.e.x.t
+	 * @var Context
+	 */
+	private $context;
+
+	/**
+	 * Modules instance.
+	 *
+	 * @since n.e.x.t
+	 * @var Modules
+	 */
+	private $modules;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Context $context Context instance.
+	 */
+	public function __construct( Context $context ) {
+		$this->context = $context;
+		$this->modules = new Modules( $context );
+	}
+
+	/**
+	 * Registers debug information with Site Health.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function register() {
+		add_filter(
+			'debug_information',
+			function ( $info ) {
+				$info['google-site-kit'] = array(
+					'label'  => __( 'Site Kit by Google', 'google-site-kit' ),
+					'fields' => $this->get_fields(),
+				);
+
+				return $info;
+			}
+		);
+	}
+
+	/**
+	 * Redacts the given string by overwriting a portion with a mask character.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $string     Input string to redact.
+	 * @param int    $mask_start Starting position of redaction and length of preserved characters.
+	 *                             If negative, characters are redacted from the beginning preserving the last X characters.
+	 *                             If positive, characters are redacted from the end, preserving the first X characters.
+	 * @return string
+	 */
+	public static function redact_debug_value( $string, $mask_start = -4 ) {
```

Comment:
Still wondering whether we should go with `4` instead. While the beginning is often the same, I'd argue seeing these characters helps us more than seeing the last 4. For example, it would allow to spot us if somebody has an entirely invalid value set, whereas seeing the last 4 wouldn't likely help us at all, because the only way it would to identify would be with the entire value (which we shouldn't do as discussed). E.g. if someone has `1234` instead of `UA-x` for GA property, we know something is wrong - if we had the last 4 characters, we wouldn't know that.

Reply:
That's a good point, I see what you mean. Perhaps we should make the unmasked portion a few characters longer then to account for the characters that are used by letters that will always be the same?

Reply:
Makes sense, the lengths you chose look good to me.